### PR TITLE
feat(plugin): implement golang version of plugin jwt-auth

### PR DESCRIPTION
**Pull request by Ink-33**
### Ⅰ. Describe what this PR did
This merge request contains the `jwt-auth` plugin implemented in golang.   

### Ⅱ. Does this pull request fix one issue?

#232 , #591 

### Ⅲ. Why don't you add test cases (unit test/integration test)? 
Test cases are still in progress due to the complexity.

### Ⅳ. Describe how to verify it
As #589, all behaviors of this plugin should be same with the currently cpp verison of `jwt-auth`. Maybe we need to use plently of tese cases to verify it.

### Ⅴ. Special notes for reviews
Same with IV. Since I am not particularly familiar with cpp, I hope that reviewers will understand some errors and I will correct them in time🙏

### CODE REVIEW

**Comment by johnlanni** [comment]
@Ink-33 Is this draft PR ready for review, or is there still unfinished work?

**Comment by Ink-33** [comment]
> @Ink-33 Is this draft PR ready for review, or is there still unfinished work?

Yes it's ready for review and I will update the branch later today.

**Comment by johnlanni** [comment]
@Ink-33 Please add an e2e test for it.

File: plugins/wasm-go/extensions/jwt-auth/go.mod [hunk 1]
```diff
@@ -0,0 +1,24 @@
+module github.com/alibaba/higress/plugins/wasm-go/extensions/jwt-auth
+
+go 1.21
+
+toolchain go1.21.5
```
Comment (Ink-33) [inline plugins/wasm-go/extensions/jwt-auth/go.mod:5]: should be removed

File: plugins/wasm-go/extensions/jwt-auth/go.mod [hunk 2]
```diff
@@ -0,0 +1,24 @@
+module github.com/alibaba/higress/plugins/wasm-go/extensions/jwt-auth
+
+go 1.21
```
Comment (Ink-33) [inline plugins/wasm-go/extensions/jwt-auth/go.mod:3]: switch to 1.19

**Comment by johnlanni** [comment]
@Ink-33 Please add e2e test cases, refer to: https://github.com/alibaba/higress/pull/759/files#diff-00090f4cf6417f2566ebc91e56b6b5d0e12a834a184f4b814d51bc89581d796b

**Comment by Ink-33** [comment]
> @Ink-33 Please add e2e test cases, refer to: https://github.com/alibaba/higress/pull/759/files#diff-00090f4cf6417f2566ebc91e56b6b5d0e12a834a184f4b814d51bc89581d796b

got it. I would like to add three parts of test cases including normal, rejected, and normal with all features one by one.

btw, how to run e2e test case on my local dev environments?

**Comment by johnlanni** [comment]
> > @Ink-33 Please add e2e test cases, refer to: https://github.com/alibaba/higress/pull/759/files#diff-00090f4cf6417f2566ebc91e56b6b5d0e12a834a184f4b814d51bc89581d796b
> 
> got it. I would like to add three parts of test cases including normal, rejected, and normal with all features one by one.
> 
> btw, how to run e2e test case on my local dev environments?

Please refer this guide: https://github.com/alibaba/higress/tree/main/test#how-to-implement-test-environment-reusability

File: test/e2e/conformance/tests/go-wasm-jwt-auth.go [hunk 3]
```diff
@@ -0,0 +1,703 @@
...
+
+package tests
+
+import (
+	"testing"
+
+	"github.com/alibaba/higress/test/e2e/conformance/utils/http"
+	"github.com/alibaba/higress/test/e2e/conformance/utils/suite"
+)
+
```
Comment (Ink-33) [inline test/e2e/conformance/tests/go-wasm-jwt-auth.go:23]: here need to add
```
Register(WasmPluginsJWTAuth...)
```

**Comment by Ink-33** [comment]
@johnlanni allow 里面配置多个 consumer 且全部验证失败应该返回哪个错误码给请求，或者直接用 verify failed 兜底呢？

**Comment by johnlanni** [comment]
> @johnlanni allow 里面配置多个 consumer 且全部验证失败应该返回哪个错误码给请求，或者直接用 verify failed 兜底呢？
> 
> 

返回 403 即可，错误原因可以用 verify failed

File: plugins/wasm-go/extensions/jwt-auth/config/parser.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+			log.Warn(err.Error())
+			continue
+		}
+		config.Consumers = append(config.Consumers, c)
+	}
+	if len(config.Consumers) == 0 {
+		return fmt.Errorf("at least one consumer should be configured for a rule")
+	}
+
+	// GlobalAuth 为 true 时代表插件全局启用，需要标记 enabled 为 true
```
Comment (johnlanni) [inline plugins/wasm-go/extensions/jwt-auth/config/parser.go:51]: 貌似没看到解析生成 config.GlobalAuth的代码？

类似：https://github.com/alibaba/higress/blob/e09edff827b94fa5bcc149bbeadc905361100c2a/plugins/wasm-go/extensions/basic-auth/main.go#L149-L153v

**Review by johnlanni — CHANGES_REQUESTED** [review]

File: plugins/wasm-go/extensions/jwt-auth/config/parser.go (hunk 4 above)
Comment (Ink-33) [reply plugins/wasm-go/extensions/jwt-auth/config/parser.go:51]: 我用那个GlobalAuthCheck去做了一下判断但是好像没覆盖全部情况，我看下怎么改

**Review by johnlanni — APPROVED** [review]
LGTM, Thanks！

**Pull request merged**